pkg/auth/providers/azure/clients/ms_graph_client.go

```diff
@@ -232,27 +232,72 @@ type AccessTokenCache struct {
 func (c AccessTokenCache) Replace(cache cache.Unmarshaler, key string) {
 	secret, err := common.ReadFromSecret(c.Secrets, common.SecretsNamespace+":azuread-access-token", "access-token")
 	if err != nil {
-		logrus.Errorf("failed to read the access token from the database: %v", err)
+		logrus.Errorf("[AZUREAD_PROVIDER_CACHE] failed to read the access token from the database: %v", err)
 		return
 	}
 
 	err = cache.Unmarshal([]byte(secret))
 	if err != nil {
-		logrus.Errorf("failed to put the access token secret into the cache: %v", err)
+		logrus.Errorf("[AZUREAD_PROVIDER_CACHE] failed to put the access token secret into the cache: %v", err)
 	}
 }
 
-// Export persists the access token to the secret in the database.
+// Export persists the access token to a secret in the database. Before doing that, it validates the access token
+// to ensure it contains the necessary permissions, so as not to persist a token with wrong or insufficient permissions.
 func (c AccessTokenCache) Export(cache cache.Marshaler, key string) {
 	marshalled, err := cache.Marshal()
 	if err != nil {
-		logrus.Errorf("failed to marshal the access token before saving to the database: %v", err)
+		logrus.Errorf("[AZUREAD_PROVIDER_CACHE] failed to marshal the access token before saving to the database: %v", err)
 		return
 	}
+
+	// This will fail once when the caching mechanism initially tries to find the value before Azure AD is fully set up.
+	token, err := extractAccessTokenFromSecret(marshalled)
+	if err != nil {
+		logrus.Errorf("[AZUREAD_PROVIDER_CACHE] failed to extract the Microsoft Graph access token for validation: %v", err)
+		return
+	}
+	if err := EnsureMSGraphTokenHasPermissions(token); err != nil {
+		logrus.Errorf("[AZUREAD_PROVIDER_CACHE] error validating the Microsoft Graph access token: %v", err)
+		return
+	}
+
 	err = common.CreateOrUpdateSecrets(c.Secrets, string(marshalled), "access-token", "azuread")
 	if err != nil {
-		logrus.Errorf("failed to save the access token in the database: %v", err)
+		logrus.Errorf("[AZUREAD_PROVIDER_CACHE] failed to save the access token in the database: %v", err)
+	}
+}
+
+// extractAccessTokenFromSecret tries to find the JWT with access to the Microsoft Graph API in the JSON value stored
+// in a secret in the database.
+func extractAccessTokenFromSecret(secret []byte) (string, error) {
```

Comment:
This method seems quite complex. Is there any way that we can give the raw value to an already imported third party library and have it do the parsing of it for us, rather than doing this parsing ourselves as below?

Reply:
Yes, I agree, I wonder if there is a third-party library that allows to pass a string for the sequence of keys, how one can do with jq or jsonpath.